airflow/utils/dag_processing.py

```diff
@@ -543,94 +543,86 @@ def wait_until_finished(self):
+        # Reload configurations and settings to avoid collision with parent process.
+        # Because this process may need custom configurations that cannot be shared,
+        # e.g. RotatingFileHandler. And it can cause connection corruption if we
+        # do not recreate the SQLA connection pool.
+        os.environ['CONFIG_PROCESSOR_MANAGER_LOGGER'] = 'True'
+        # Replicating the behavior of how logging module was loaded
+        # in logging_config.py
+        reload_module(import_module(airflow.settings.LOGGING_CLASS_PATH.rsplit('.', 1)[0]))
+        reload_module(airflow.settings)
+        airflow.settings.initialize()
+        del os.environ['CONFIG_PROCESSOR_MANAGER_LOGGER']
+        processor_manager = DagFileProcessorManager(dag_directory,
+                                                    file_paths,
+                                                    max_runs,
+                                                    processor_factory,
+                                                    signal_conn,
+                                                    async_mode)
+
+        processor_manager.start()
 
     def harvest_simple_dags(self):
         """
         Harvest DAG parsing results from result queue and sync metadata from stat queue.
+
         :return: List of parsing result in SimpleDag format.
         """
-        # Metadata and results to be harvested can be inconsistent,
-        # but it should not be a big problem.
-        self._sync_metadata()
-        # Heartbeating after syncing metadata so we do not restart manager
-        # if it processed all files for max_run times and exit normally.
-        self._heartbeat_manager()
+
         simple_dags = []
-        while True:
-            try:
-                result = self._result_queue.get_nowait()
-                try:
-                    simple_dags.append(result)
-                finally:
-                    self._result_queue.task_done()
-            except Empty:
-                break
+        while self._parent_signal_conn.poll():
+            result = self._parent_signal_conn.recv()
+
+            self.log.debug("Received message of type %s", type(result).__name__)
+            if isinstance(result, DagParsingStat):
+                self._sync_metadata(result)
+            else:
+                simple_dags.append(result)
+
+        # Receive any pending messages before checking if the process has exited.
+        self._heartbeat_manager()
 
         self._result_count = 0
 
         return simple_dags
 
     def _heartbeat_manager(self):
         """
-        Heartbeat DAG file processor and start it if it is not alive.
-        :return:
+        Heartbeat DAG file processor and retstart it if we are not done.
         """
-        if self._process and not self._process.is_alive() and not self.done:
-            self.start()
+        if self._process and not self._process.is_alive():
+            self._process.join(timeout=0)
```

Comment:
Again here - do we have to join() here if process is not alive() ?

Reply:
Based on the code in Py3.5 yes I think we do :https://github.com/python/cpython/blob/3.5/Lib/multiprocessing/process.py#L118-L138 - without joining the process we never remove the process from the `_children` internal state in the MP module.

Reply:
OK. good